File: python/examples/fsm/01-light/light_fsm.py
"""Light FSM — simple toggle.

One button, one light.  Button press toggles ON ↔ OFF.
Equivalent to ``c/examples/fsm/00-light/light_fsm.c``.
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[4]))

from rxnet.fsm import Machine, Transition
from rxnet.runtime import Context

# Import shared driver relative to the examples root.
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
import app_driver

LIGHT_STATE_OFF = 0
LIGHT_STATE_ON = 1


class _Data:
    __slots__ = ("button_gpio", "light_gpio", "latched_event", "event_consumed", "output_enabled")

    def __init__(self, button_gpio: int, light_gpio: int) -> None:
        self.button_gpio = button_gpio
        self.light_gpio = light_gpio
        self.latched_event = False
        self.event_consumed = False
        self.output_enabled = False


def _latch(ctx: Context, data: _Data) -> None:
    data.latched_event = app_driver.latch_button_event(data.button_gpio)
    data.event_consumed = False


def _button_pressed(ctx: Context, data: _Data) -> bool:
    return data.latched_event


def _light_on(ctx: Context, data: _Data) -> None:
    data.output_enabled = True
    data.event_consumed = True


def _light_off(ctx: Context, data: _Data) -> None:
    data.output_enabled = False
    data.event_consumed = True


def _dump(ctx: Context, data: _Data) -> None:
    app_driver.set_light(data.light_gpio, data.output_enabled)
    if data.event_consumed:
        app_driver.clear_button_event(data.button_gpio)


_TRANSITIONS = [
    Transition(LIGHT_STATE_OFF, LIGHT_STATE_ON, guard=_button_pressed, action=_light_on),
    Transition(LIGHT_STATE_ON, LIGHT_STATE_OFF, guard=_button_pressed, action=_light_off),
]


def create_light_fsm(button_gpio: int, light_gpio: int) -> Machine:
    app_driver.init_button(button_gpio)
    app_driver.init_light(light_gpio)
    data = _Data(button_gpio, light_gpio)
    return Machine(
        name="light",
        state=LIGHT_STATE_OFF,
        transitions=_TRANSITIONS,
        user=data,
        latch_inputs_cb=_latch,
        dump_outputs_cb=_dump,
    )

File: c/examples/fsm/01-light/light_fsm.c
// Copyright 2026 Jose M. Moya <jm.moya@upm.es>
// SPDX-License-Identifier: MIT

#include "light_fsm.h"

#include <stddef.h>
#include <string.h>

#include "rxnet/config.h"

typedef struct {
    bool in_use;
    rx_fsm_machine *machine;
    gpio_num_t button_gpio;
    gpio_num_t light_gpio;
    bool latched_event;
    bool event_consumed;
    int output_enabled;
} light_machine_data;

static light_machine_data s_machine_data[RXNET_MAX_RUNTIME_NODES];

static light_machine_data *find_or_allocate_data(rx_fsm_machine *machine) {
    size_t i;

    for (i = 0; i < RXNET_MAX_RUNTIME_NODES; ++i) {
        if (s_machine_data[i].in_use && s_machine_data[i].machine == machine) {
            return &s_machine_data[i];
        }
    }

    for (i = 0; i < RXNET_MAX_RUNTIME_NODES; ++i) {
        if (!s_machine_data[i].in_use) {
            memset(&s_machine_data[i], 0, sizeof(s_machine_data[i]));
            s_machine_data[i].in_use = true;
            s_machine_data[i].machine = machine;
            return &s_machine_data[i];
        }
    }

    return NULL;
}

enum {
    LIGHT_STATE_OFF = 0,
    LIGHT_STATE_ON = 1,
};

static void light_machine_latch_inputs(rx_fsm_context *ctx, void *user) {
    light_machine_data *data = (light_machine_data *)user;

    if (data == NULL) {
        return;
    }

    (void)ctx;
    data->latched_event = app_driver_latch_button_event(data->button_gpio);
    data->event_consumed = false;
}

static int button_pressed(const rx_fsm_context *ctx, void *user) {
    const light_machine_data *data = (const light_machine_data *)user;

    return data != NULL && data->latched_event;
}

static void light_on(rx_fsm_context *ctx, void *user) {
    light_machine_data *data = (light_machine_data *)user;
    (void)ctx;
    if (data == NULL) {
        return;
    }
    data->output_enabled = 1;
    data->event_consumed = true;
}

static void light_off(rx_fsm_context *ctx, void *user) {
    light_machine_data *data = (light_machine_data *)user;
    (void)ctx;
    if (data == NULL) {
        return;
    }
    data->output_enabled = 0;
    data->event_consumed = true;
}

static void light_machine_dump_outputs(rx_fsm_context *ctx, void *user) {
    light_machine_data *data = (light_machine_data *)user;

    (void)ctx;
    if (data == NULL) {
        return;
    }

    app_driver_set_light(data->light_gpio, data->output_enabled);
    if (data->event_consumed) {
        app_driver_clear_button_event(data->button_gpio);
    }
}

void light_fsm_create(
    rx_fsm_machine *machine,
    gpio_num_t button_gpio,
    gpio_num_t light_gpio
) {
    static const rx_fsm_transition transitions[] = {
        {LIGHT_STATE_OFF, LIGHT_STATE_ON, button_pressed, light_on},
        {LIGHT_STATE_ON, LIGHT_STATE_OFF, button_pressed, light_off},
    };
    light_machine_data *data = find_or_allocate_data(machine);

    if (machine == NULL || data == NULL) {
        return;
    }

    data->button_gpio = button_gpio;
    data->light_gpio = light_gpio;
    data->latched_event = false;
    data->event_consumed = false;
    data->output_enabled = 0;

    if (app_driver_init_button(button_gpio) != ESP_OK ||
        app_driver_init_light(light_gpio) != ESP_OK) {
        data->in_use = false;
        return;
    }

    rx_fsm_machine_init(
        machine,
        "light",
        LIGHT_STATE_OFF,
        transitions,
        sizeof(transitions) / sizeof(transitions[0]),
        data,
        light_machine_latch_inputs,
        light_machine_dump_outputs
    );
}
